File: kernel/src/domain/comments/comments.ts
/**
 * Comments Domain Module
 *
 * Delegates all data access to ComputeBridge (Rust compute-core).
 *
 * Architecture:
 * - Write operations: fire-and-forget via ctx.computeBridge
 * - Read operations: async via ctx.computeBridge
 * - Events: handled by MutationResultHandler from Rust MutationResult
 * - Orphan cleanup: handled by Rust during structure operations
 *
 * Cell Identity Model:
 * - Comments reference cells via CellId (stable UUID)
 * - Position resolved at render time via GridIndex.getPosition()
 *
 * @see compute-core/src/storage/comments.rs - Rust implementation
 */

import { toCellId, type CellId } from '@mog-sdk/contracts/cell-identity';
import type { Comment } from '@mog-sdk/contracts/comments';
import type { SheetId } from '@mog-sdk/contracts/core';
import type { StructureChangeSource } from '@mog-sdk/contracts/event-base';
import type { RichText } from '@mog-sdk/contracts/rich-text';

import type { Comment as BridgeComment } from '../../bridges/compute/compute-types.gen';
import type { DocumentContext } from '../../context/types';

/**
 * Map a bridge Comment (from Rust) to a contracts Comment (for UI).
 * Bridge has `runs: RichTextRun[]` + `content?: string`.
 * Contracts has `content: RichText` (array of segments with TextFormat).
 */
function fromBridgeComment(bc: BridgeComment): Comment {
  return {
    id: bc.id,
    cellRef: toCellId(bc.cellRef),
    author: bc.author,
    authorId: bc.authorId,
    createdAt: bc.createdAt ?? 0,
    modifiedAt: bc.modifiedAt ?? undefined,
    content: bc.runs.map((r) => ({
      text: r.text,
      format:
        r.bold || r.italic || r.underline || r.strikethrough || r.color
          ? {
              bold: r.bold || undefined,
              italic: r.italic || undefined,
              underlineType: r.underline ? ('single' as const) : undefined,
              strikethrough: r.strikethrough || undefined,
              fontColor: r.color ?? undefined,
            }
          : undefined,
    })),
    threadId: bc.threadId || undefined,
    parentId: bc.parentId ?? undefined,
    resolved: bc.resolved ?? undefined,
    commentType: bc.commentType,
  };
}

// =============================================================================
// Read Operations
// =============================================================================

/**
 * Get all comments for a cell.
 * Comments are sorted by createdAt timestamp.
 *
 * @param ctx - Store context
 * @param sheetId - Sheet ID
 * @param cellId - CellId to get comments for
 * @returns Promise of comments array, sorted by creation time
 */
export async function getCommentsForCell(
  ctx: DocumentContext,
  sheetId: SheetId,
  cellId: CellId,
): Promise<Comment[]> {
  const bridge = await ctx.computeBridge.getCommentsForCell(sheetId, cellId);
  return bridge.map(fromBridgeComment);
}

/**
 * Get a single comment by ID.
 *
 * @param ctx - Store context
 * @param sheetId - Sheet ID
 * @param commentId - Comment ID
 * @returns The comment, or undefined if not found
 */
export async function getComment(
  ctx: DocumentContext,
  sheetId: SheetId,
  commentId: string,
): Promise<Comment | undefined> {
  const comment = await ctx.computeBridge.getComment(sheetId, commentId);
  return comment ? fromBridgeComment(comment) : undefined;
}

/**
 * Check if a cell has any comments.
 *
 * @param ctx - Store context
 * @param sheetId - Sheet ID
 * @param cellId - CellId to check
 * @returns True if cell has at least one comment
 */
export async function hasComments(
  ctx: DocumentContext,
  sheetId: SheetId,
  cellId: CellId,
): Promise<boolean> {
  const comments = await ctx.computeBridge.getCommentsForCell(sheetId, cellId);
  return comments.length > 0;
}

/**
 * Get all comments in a sheet.
 *
 * @param ctx - Store context
 * @param sheetId - Sheet ID
 * @returns Promise of all comments in the sheet
 */
export async function getAll(ctx: DocumentContext, sheetId: SheetId): Promise<Comment[]> {
  const bridge = await ctx.computeBridge.getAllComments(sheetId);
  return bridge.map(fromBridgeComment);
}

/**
 * Get all CellIds that have comments (for rendering indicators).
 * This is used by the cells layer to know which cells need comment triangles.
 *
 * @param ctx - Store context
 * @param sheetId - Sheet ID
 * @returns Set of CellIds that have comments
 */
export async function getCellIdsWithComments(
  ctx: DocumentContext,
  sheetId: SheetId,
): Promise<Set<CellId>> {
  const all = await ctx.computeBridge.getAllComments(sheetId);
  const cellIds = new Set<CellId>();
  for (const comment of all) {
    cellIds.add(toCellId(comment.cellRef));
  }
  return cellIds;
}

/**
 * Get the count of comments in a sheet.
 *
 * @param ctx - Store context
 * @param sheetId - Sheet ID
 * @returns Number of comments in the sheet
 */
export async function getCount(ctx: DocumentContext, sheetId: SheetId): Promise<number> {
  const all = await ctx.computeBridge.getAllComments(sheetId);
  return all.length;
}

/**
 * Get all comments in a thread.
 * A thread is identified by its root comment's ID (threadId).
 *
 * @param ctx - Store context
 * @param sheetId - Sheet ID
 * @param threadId - ID of the thread root comment
 * @returns Array of comments in the thread, sorted by creation time
 */
export async function getThread(
  ctx: DocumentContext,
  sheetId: SheetId,
  threadId: string,
): Promise<Comment[]> {
  const bridge = await ctx.computeBridge.getCommentThread(sheetId, threadId);
  return bridge.map(fromBridgeComment);
}

// =============================================================================
// Write Operations
// =============================================================================

/**
 * Add a new comment to a cell.
 * Delegates to ComputeBridge; events emitted via MutationResultHandler.
 *
 * @param ctx - Store context
 * @param sheetId - Sheet ID
 * @param cellId - CellId to attach comment to
 * @param content - Rich text content
 * @param author - Author display name
 * @param options - Optional: authorId, parentId for replies (handled by Rust)
 * @param _source - Origin of the change (handled by Rust)
 */
export async function addComment(
  ctx: DocumentContext,
  sheetId: SheetId,
  cellId: CellId,
  content: RichText,
  author: string,
  options?: { authorId?: string; parentId?: string },
  _source: StructureChangeSource = 'user',
): Promise<Comment> {
  // Convert RichText to plain text for the bridge
  const text = content.map((segment) => segment.text).join('');
  const bridge = await ctx.computeBridge.addComment(sheetId, cellId, text, author, options);
  return fromBridgeComment(bridge);
}

/**
 * Update a comment's content.
 * Delegates to ComputeBridge; events emitted via MutationResultHandler.
 *
 * @param ctx - Store context
 * @param sheetId - Sheet ID
 * @param commentId - Comment ID to update
 * @param content - New rich text content
 * @param _source - Origin of the change (handled by Rust)
 */
export function updateComment(
  ctx: DocumentContext,
  sheetId: SheetId,
  commentId: string,
  content: RichText,
  _source: StructureChangeSource = 'user',
): void {
  const text = content.map((segment) => segment.text).join('');
  void ctx.computeBridge.updateComment(sheetId, commentId, text);
}

/**
 * Delete a comment.
 * Delegates to ComputeBridge; events emitted via MutationResultHandler.
 *
 * @param ctx - Store context
 * @param sheetId - Sheet ID
 * @param commentId - Comment ID to delete
 * @param _source - Origin of the change (handled by Rust)
 */
export function deleteComment(
  ctx: DocumentContext,
  sheetId: SheetId,
  commentId: string,
  _source: StructureChangeSource = 'user',
): void {
  void ctx.computeBridge.deleteComment(sheetId, commentId);
}

/**
 * Delete all comments for a cell.
 * Fetches comments for the cell, then deletes each via ComputeBridge.
 *
 * @param ctx - Store context
 * @param sheetId - Sheet ID
 * @param cellId - CellId to delete comments from
 * @param _source - Origin of the change (handled by Rust)
 */
export function deleteCommentsForCell(
  ctx: DocumentContext,
  sheetId: SheetId,
  cellId: CellId,
  _source: StructureChangeSource = 'user',
): void {
  void (async () => {
    const comments = await ctx.computeBridge.getCommentsForCell(sheetId, cellId);
    await Promise.all(
      comments.map((comment) => ctx.computeBridge.deleteComment(sheetId, comment.id)),
    );
  })();
}

/**
 * Resolve or unresolve a thread.
 * Delegates to ComputeBridge; events emitted via MutationResultHandler.
 *
 * @param ctx - Store context
 * @param sheetId - Sheet ID
 * @param threadId - ID of the thread root comment
 * @param resolved - Whether the thread should be resolved
 * @param _source - Origin of the change (handled by Rust)
 */
export function setThreadResolved(
  ctx: DocumentContext,
  sheetId: SheetId,
  threadId: string,
  resolved: boolean,
  _source: StructureChangeSource = 'user',
): void {
  void ctx.computeBridge.setThreadResolved(sheetId, threadId, resolved);
}

/**
 * Clear all comments for a sheet.
 * Fetches all comments, then deletes each via ComputeBridge.
 *
 * @param ctx - Store context
 * @param sheetId - Sheet ID
 * @param _source - Origin of the change (handled by Rust)
 */
export function clearAll(
  ctx: DocumentContext,
  sheetId: SheetId,
  _source: StructureChangeSource = 'user',
): void {
  void (async () => {
    const comments = await ctx.computeBridge.getAllComments(sheetId);
    await Promise.all(
      comments.map((comment) => ctx.computeBridge.deleteComment(sheetId, comment.id)),
    );
  })();
}

// =============================================================================
// Validation Helpers (for orphaned comments)
// =============================================================================

/**
 * Validate and clean orphaned comments.
 *
 * In the ComputeBridge architecture, orphan cleanup is handled by Rust
 * during structure operations (deleteRows, deleteCols). This function
 * is a no-op stub kept for API compatibility with callers in structures.ts.
 *
 * @param _ctx - Store context (unused)
 * @param _sheetId - Sheet ID (unused)
 * @param _source - Origin of the change (unused)
 * @returns Always 0 (cleanup happens in Rust)
 */
export function validateAndClean(
  _ctx: DocumentContext,
  _sheetId: SheetId,
  _source: StructureChangeSource = 'user',
): number {
  // Orphan cleanup is handled by Rust compute-core during structure operations.
  // See compute-core/src/storage/comments.rs: validate_and_clean_comments()
  return 0;
}

// =============================================================================
// Subscribe to Comments
// =============================================================================

/**
 * Subscribe to comment changes for a sheet.
 * The callback receives all comments when any change occurs.
 *
 * NOTE: In the ComputeBridge architecture, subscriptions should be wired
 * through the MutationResult event system. This stub returns a no-op
 * unsubscribe for API compatibility.
 *
 * @param _ctx - Store context
 * @param _sheetId - Sheet ID
 * @param _callback - Called when comments change
 * @returns Unsubscribe function (no-op)
 */
export function subscribe(
  _ctx: DocumentContext,
  _sheetId: SheetId,
  _callback: (comments: Comment[]) => void,
): () => void {
  // In the ComputeBridge architecture, comment change notifications come
  // through MutationResult events, not CRDT observe. This is a compatibility stub.
  return () => {};
}

File: compute/api/src/sheet/comments.rs
//! Threaded comment operations for a sheet.

use crate::dispatch::Dispatch;
use crate::error::ComputeApiError;
use cell_types::SheetId;
use domain_types::Comment;
use domain_types::domain::comment::{CommentMention, CommentType};
use snapshot_types::MutationResult;

/// Sub-API for comment operations on a single sheet.
pub struct SheetComments {
    dispatch: Dispatch,
    sheet_id: SheetId,
}

impl SheetComments {
    pub(crate) fn new(dispatch: Dispatch, sheet_id: SheetId) -> Self {
        Self { dispatch, sheet_id }
    }

    // -----------------------------------------------------------------
    // Queries
    // -----------------------------------------------------------------

    /// Get all comments attached to a specific cell.
    pub fn get_for_cell(&self, cell_id: &str) -> Result<Vec<Comment>, ComputeApiError> {
        let sid = self.sheet_id;
        let cid = cell_id.to_string();
        self.dispatch
            .query_engine(move |e| e.get_comments_for_cell(&sid, &cid))
    }

    /// Get all comments in this sheet.
    pub fn get_all(&self) -> Result<Vec<Comment>, ComputeApiError> {
        let sid = self.sheet_id;
        self.dispatch
            .query_engine(move |e| e.get_all_comments(&sid))
    }

    /// Get a single comment by its ID.
    pub fn get(&self, comment_id: &str) -> Result<Option<Comment>, ComputeApiError> {
        let sid = self.sheet_id;
        let cid = comment_id.to_string();
        self.dispatch
            .query_engine(move |e| e.get_comment(&sid, &cid))
    }

    /// Get all comments in a thread, sorted by creation time.
    pub fn get_thread(&self, thread_id: &str) -> Result<Vec<Comment>, ComputeApiError> {
        let sid = self.sheet_id;
        let tid = thread_id.to_string();
        self.dispatch
            .query_engine(move |e| e.get_comment_thread(&sid, &tid))
    }

    /// Get the total number of comments in this sheet.
    pub fn count(&self) -> Result<u32, ComputeApiError> {
        let sid = self.sheet_id;
        self.dispatch
            .query_engine(move |e| e.get_comment_count(&sid))
    }

    /// Get the count of notes (comments with `comment_type == Note`) in this sheet.
    pub fn note_count(&self) -> Result<u32, ComputeApiError> {
        let sid = self.sheet_id;
        self.dispatch.query_engine(move |e| e.get_note_count(&sid))
    }

    /// Get all notes (comments with `comment_type == Note`) in this sheet.
    pub fn get_all_notes(&self) -> Result<Vec<Comment>, ComputeApiError> {
        let sid = self.sheet_id;
        self.dispatch.query_engine(move |e| e.get_all_notes(&sid))
    }

    /// Set the `visible` flag on a note (VML note visibility).
    pub fn set_note_visible(
        &self,
        comment_id: &str,
        visible: bool,
    ) -> Result<MutationResult, ComputeApiError> {
        let sid = self.sheet_id;
        let cid = comment_id.to_string();
        self.dispatch
            .call_engine(move |e| e.set_note_visible(&sid, &cid, visible))
            .and_then(|r| {
                r.map(|(_patches, result)| result)
                    .map_err(ComputeApiError::from)
            })
    }

    /// Set the height and/or width of a note (in points).
    pub fn set_note_dimensions(
        &self,
        comment_id: &str,
        height: Option<f64>,
        width: Option<f64>,
    ) -> Result<MutationResult, ComputeApiError> {
        let sid = self.sheet_id;
        let cid = comment_id.to_string();
        self.dispatch
            .call_engine(move |e| e.set_note_dimensions(&sid, &cid, height, width))
            .and_then(|r| {
                r.map(|(_patches, result)| result)
                    .map_err(ComputeApiError::from)
            })
    }

    /// Check whether a cell has any comments.
    pub fn has_comments(&self, cell_id: &str) -> Result<bool, ComputeApiError> {
        let sid = self.sheet_id;
        let cid = cell_id.to_string();
        self.dispatch
            .query_engine(move |e| e.has_comments(&sid, &cid))
    }

    // -----------------------------------------------------------------
    // Mutations
    // -----------------------------------------------------------------

    /// Add a comment to a cell.
    ///
    /// `comment_type` is the discriminator: `Note` writes a legacy note (no
    /// thread membership, can't have replies), `ThreadedComment` writes a
    /// modern threaded comment. Notes with a non-`None` `parent_id` are
    /// rejected as a contract violation.
    #[allow(clippy::too_many_arguments)]
    pub fn add(
        &self,
        cell_id: &str,
        author: &str,
        text: &str,
        author_id: Option<&str>,
        parent_id: Option<&str>,
        comment_type: CommentType,
    ) -> Result<MutationResult, ComputeApiError> {
        let sid = self.sheet_id;
        let cid = cell_id.to_string();
        let auth = author.to_string();
        let txt = text.to_string();
        let aid = author_id.map(|s| s.to_string());
        let pid = parent_id.map(|s| s.to_string());
        self.dispatch
            .call_engine(move |e| e.add_comment(&sid, &cid, &txt, &auth, aid, pid, comment_type))
            .and_then(|r| {
                r.map(|(_patches, result)| result)
                    .map_err(ComputeApiError::from)
            })
    }

    /// Convert an existing note into a threaded comment. Returns
    /// `MutationResult` whose `data` contains the updated `Comment`.
    pub fn convert_note_to_thread(
        &self,
        comment_id: &str,
    ) -> Result<MutationResult, ComputeApiError> {
        let sid = self.sheet_id;
        let cid = comment_id.to_string();
        self.dispatch
            .call_engine(move |e| e.convert_note_to_thread(&sid, &cid))
            .and_then(|r| {
                r.map(|(_patches, result)| result)
                    .map_err(ComputeApiError::from)
            })
    }

    /// Update a comment's text.
    pub fn update(&self, comment_id: &str, text: &str) -> Result<MutationResult, ComputeApiError> {
        let sid = self.sheet_id;
        let cid = comment_id.to_string();
        let txt = text.to_string();
        self.dispatch
            .call_engine(move |e| e.update_comment(&sid, &cid, &txt))
            .and_then(|r| {
                r.map(|(_patches, result)| result)
                    .map_err(ComputeApiError::from)
            })
    }

    /// Update a comment with mention content.
    pub fn update_mentions(
        &self,
        comment_id: &str,
        content: &str,
        mentions: Vec<CommentMention>,
    ) -> Result<MutationResult, ComputeApiError> {
        let sid = self.sheet_id;
        let cid = comment_id.to_string();
        let cnt = content.to_string();
        self.dispatch
            .call_engine(move |e| e.update_comment_mentions(&sid, &cid, &cnt, mentions))
            .and_then(|r| {
                r.map(|(_patches, result)| result)
                    .map_err(ComputeApiError::from)
            })
    }

    /// Delete a comment by ID.
    pub fn delete(&self, comment_id: &str) -> Result<MutationResult, ComputeApiError> {
        let sid = self.sheet_id;
        let cid = comment_id.to_string();
        self.dispatch
            .call_engine(move |e| e.delete_comment(&sid, &cid))
            .and_then(|r| {
                r.map(|(_patches, result)| result)
                    .map_err(ComputeApiError::from)
            })
    }

    /// Set thread resolved status.
    pub fn set_thread_resolved(
        &self,
        cell_id: &str,
        resolved: bool,
    ) -> Result<MutationResult, ComputeApiError> {
        let sid = self.sheet_id;
        let cid = cell_id.to_string();
        self.dispatch
            .call_engine(move |e| e.set_thread_resolved(&sid, &cid, resolved))
            .and_then(|r| {
                r.map(|(_patches, result)| result)
                    .map_err(ComputeApiError::from)
            })
    }

    /// Delete all comments associated with a specific cell.
    pub fn delete_for_cell(&self, cell_id: &str) -> Result<MutationResult, ComputeApiError> {
        let sid = self.sheet_id;
        let cid = cell_id.to_string();
        self.dispatch
            .call_engine(move |e| e.delete_comments_for_cell(&sid, &cid))
            .and_then(|r| {
                r.map(|(_patches, result)| result)
                    .map_err(ComputeApiError::from)
            })
    }

    /// Remove all comments from this sheet.
    pub fn clear_all(&self) -> Result<MutationResult, ComputeApiError> {
        let sid = self.sheet_id;
        self.dispatch
            .call_engine(move |e| e.clear_all_comments(&sid))
            .and_then(|r| {
                r.map(|(_patches, result)| result)
                    .map_err(ComputeApiError::from)
            })
    }
}
